dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/jsonpatch/JsonPatchManagerTest.java

```diff
@@ -0,0 +1,142 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.jsonpatch;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import org.hisp.dhis.IntegrationTestBase;
+import org.hisp.dhis.common.CodeGenerator;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.commons.jackson.config.JacksonObjectMapperConfig;
+import org.hisp.dhis.commons.jackson.jsonpatch.JsonPatch;
+import org.hisp.dhis.constant.Constant;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dataelement.DataElementGroup;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Morten Olav Hansen
+ */
+public class JsonPatchManagerTest
+    extends IntegrationTestBase
+{
+    private final ObjectMapper jsonMapper = JacksonObjectMapperConfig.staticJsonMapper();
+
+    @Autowired
+    private JsonPatchManager jsonPatchManager;
+
+    @Autowired
+    private IdentifiableObjectManager manager;
+
+    @Test
+    public void testSimpleAddPatchNoPersist()
+        throws Exception
+    {
+        Constant constant = createConstant( 'A', 1.0d );
+        assertEquals( "ConstantA", constant.getName() );
+        assertEquals( 1.0d, constant.getValue(), 0 );
+
+        JsonPatch patch = jsonMapper.readValue( "[" +
+            "{\"op\": \"add\", \"path\": \"/name\", \"value\": \"updated\"}," +
+            "{\"op\": \"add\", \"path\": \"/value\", \"value\": 5.0}" +
+            "]", JsonPatch.class );
+
+        assertNotNull( patch );
+
+        Constant patchedConstant = jsonPatchManager.apply( patch, constant );
+
+        assertEquals( "ConstantA", constant.getName() );
+        assertEquals( 1.0d, constant.getValue(), 0 );
+
+        assertEquals( "updated", patchedConstant.getName() );
+        assertEquals( 5.0d, patchedConstant.getValue(), 0 );
+    }
+
+    @Test
+    public void testSimpleAddPatchAddPersist()
+        throws Exception
+    {
+        Constant constant = createConstant( 'A', 1.0d );
+        manager.save( constant );
+
+        assertEquals( "ConstantA", constant.getName() );
+        assertEquals( 1.0d, constant.getValue(), 0 );
+
+        JsonPatch patch = jsonMapper.readValue( "[" +
+            "{\"op\": \"add\", \"path\": \"/name\", \"value\": \"updated\"}," +
+            "{\"op\": \"add\", \"path\": \"/value\", \"value\": 5.0}" +
+            "]", JsonPatch.class );
+
+        assertNotNull( patch );
+
+        Constant patchedConstant = jsonPatchManager.apply( patch, constant );
+        patchedConstant.setUid( CodeGenerator.generateUid() );
+        manager.save( patchedConstant );
+
+        assertEquals( "ConstantA", constant.getName() );
+        assertEquals( 1.0d, constant.getValue(), 0 );
+
+        assertEquals( "updated", patchedConstant.getName() );
+        assertEquals( 5.0d, patchedConstant.getValue(), 0 );
```

Comment:
Wouldn't the full test also require to (re)load the constant to see it actually got stored as patched? otherwise I don't see how this test is different from one that isn't saving the change.

Reply:
Hi, the `JsonPatchManager` does not change anything in the persisted object. What is basically does is to take in a object that it converts to a json tree, then it patches it, and returns a normal java object from it.

Persistence happens in AbstractCrudController, and is just based on our usual import. So in reality, all the patch does is to "patch" a json payload, and then run through the import process (this allows us to make any kind of change we want), the old solution patched the persisted object directly, but this had a lot of bad side effects (reloading of collections etc).

Reply:
I got that, I just wonder why you would call `save` in the test if you never verify that it had the desired effect. If it is not checked because this isn't part of the tested code the test scenario may skip `save` as well.